lib/beaker/hypervisor/aws_sdk.rb

```diff
@@ -220,125 +221,176 @@ def kill_zombie_volumes
           }
-        end
-
-        security_group = ensure_group(vpc || region, Beaker::EC2Helper.amiports(host))
-
-        # Launch the node, filling in the blanks from previous work.
-        @logger.notify("aws-sdk: Launch instance")
-        config = {
-          :count => 1,
-          :image_id => image_id,
-          :monitoring_enabled => true,
-          :key_pair => ensure_key_pair(region),
-          :security_groups => [security_group],
-          :instance_type => amisize,
-          :disable_api_termination => false,
-          :instance_initiated_shutdown_behavior => "terminate",
-          :block_device_mappings => block_device_mappings,
-          :subnet => subnet_id,
         }
-        instance = region.instances.create(config)
+      end
 
-        # Persist the instance object for this host, so later it can be
-        # manipulated by 'cleanup' for example.
-        host['instance'] = instance
+      security_group = ensure_group(vpc || region, Beaker::EC2Helper.amiports(host))
 
-        @logger.notify("aws-sdk: Launched #{host.name} (#{amitype}:#{amisize}) using snapshot/image_type #{image_type}")
-      end
+      # Launch the node, filling in the blanks from previous work.
+      @logger.notify("aws-sdk: Launch instance")
+      config = {
+        :count => 1,
+        :image_id => image_id,
+        :monitoring_enabled => true,
+        :key_pair => ensure_key_pair(region),
+        :security_groups => [security_group],
+        :instance_type => amisize,
+        :disable_api_termination => false,
+        :instance_initiated_shutdown_behavior => "terminate",
+        :block_device_mappings => block_device_mappings,
+        :subnet => subnet_id,
+      }
+
+      {instance: region.instances.create(config),
+       amitype: amitype,
+       amisize: amisize,
+       image_type: image_type}
+    end
 
+    # Create EC2 instances for all hosts, tag them, and wait until
+    # they're running.
+    #
+    # @return [void]
+    # @api private
+    def launch_all_nodes
+      @logger.notify("aws-sdk: Iterate across all hosts in configuration and launch them")
+      # Q: Right now this code looks first for a 'subnets' list
+      # (intended to be specified in CONFIG), and then, if that's not
+      # found, for the previous per-host subnet_id settings.  However,
+      # the code only supports one subnet for all hosts, so should we
+      # make it stricter, i.e. should it check that any subnets or
+      # subnet_id values match across all hosts?
+      #
+      # Q: is this the correct way to look for something that's
+      # supposed to be in CONFIG?
+      subnets = @hosts[0]['subnets']
+      if subnets.nil?
+        subnet = @hosts[0]['subnet_id'] || @options['subnet_id'] || nil
+        subnets = [subnet]
+      end
+      ami_spec = YAML.load_file(@options[:ec2_yaml])["AMI"]
+      instances = [] # Each element is {instance: i, host: h}
+      subnets.each do |subnet|
```

Comment:
On the 'different config per node' comment I had before, I wonder if we should invert this, and instead of having an all or nothing approach, just fall through on a per node basis. I see very little worries about having nodes launched in different subnet_id's (besides the attenuation I guess) and if we try the subnets in order, most will fall into the same subnet anyway.

This actually avoids the need to do an entire teardown for all nodes if just 1 node provision failed for example. Its an inversion of the way we're doing it now, and it gets us back to per node configuration again also.